### ISSUE TITLE
AzCopy and the case for adding a flag to stop version check

### ISSUE THREAD

**Issue by supermistag** [Azure/azure-storage-azcopy#1705]
##### Note: The version is visible when running AzCopy without any argument
10.14.0

### Which platform are you using? (ex: Windows, Mac, Linux)
Windows

### What command did you run?
azcopy

### What problem was encountered?
A delay of up to 1 second when starting azcopy, depending on location.

We have a large (18.5 Million) amount of files that we need to sync to blob storage. After struggling with too much CPU (Concurrency), too much memory (BUFFER) and following a variety of posts, I came to the approach of creating multiple jobs (https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-optimize).

By dicing up the folders using a PowerShell script to generate a batch file, I ended up with around 4500 calls to AzCopy. What I continually noticed, however, was a pause at the beginning of the AzCopy call (much like the 2-second bug #1653)

A process watch showed a network call on every startup taking around 1 sec. The destination URL was for a txt file of the latest version number available of AzCopy

To prove this delay problem, at the start of beginDetectNewVersion in root.go, I added 
`
		tStart := time.Now()
`
and at the end, I added
```
		t := time.Now()
		theMessage := fmt.Sprintf("Time to check new version %s",t.Sub(tStart))
		glcm.Info(theMessage)

```
**Running on my local machine in Auckland, New Zealand**
INFO: Time to check new version 1.0095371s

**Running on a system in Chicago**
INFO: Time to check new version 226.4649ms

**Running on an Azure VM in Central US**
INFO: Time to check new version 198.0786ms

Because the home location of the latest version.txt file is based at
https://azcopyvnextrelease.blob.core.windows.net/releasemetadata/latest_version.txt, every time you call AzCopy you will have a delay based on the latency between your location and this blob storage (latency+SSL negotiation)

That means on any job, you incur between 200ms and 1+ second delay before anything even starts copying.

All our transfers happen from Chicago, which means we have 4500 * 220ms = 990 seconds, or 16.5 minutes of the entire process just "looking for a new version"

What I did not test - if the system was locked down, how long before it timed out the connection?

After looking to see if there was a switch to turn this off, I was surprised to see other issues where this was requested and closed as won't do. Whilst I understand it may be important to show users that there is a newer version, it significantly impacts transfers.

Hopefully the above will be sufficient to convince that a switch to turn this behavior off is critical to large transfer jobs, as well as sites that have no open internet access to "check latest version"

**Comment by supermistag** [comment]
See #1175 and #1430

**Comment by mohsha-msft** [comment]
Hey @supermistag ,

Thanks for reaching out!

As you know we don't support disabling version check. Please allow me to correct your understanding on this a bit. Version check is done by an independent goroutine(thread) in parallel and not by main thread (sequentially) so it doesn't consume the amount of time you're claiming. Check [this code for your reference](https://github.com/Azure/azure-storage-azcopy/blob/fc4b103e4ed8ecf2ac3bcf5febdf330354e1ae46/cmd/root.go#L212). We'll not disable it as it is part of Microsoft products to do version check. 

If you still feel that it's still taking lot of time and wish to disable the version check yourself, you can comment out the invocation/occurrence of this function `beginDetectNewVersion()` and build your own binary by running `GOARCH=amd64 GOOS=windows go build -o azcopy_windows_amd64.exe` or `GOARCH=386 GOOS=windows go build -o azcopy_windows_386.exe`.

As for the [issue 1653](https://github.com/Azure/azure-storage-azcopy/issues/1653), we have fixed it and are planning to release the fix in next release.

I hope that answers your question. Please reach out for more information!

**Comment by supermistag** [comment]
Hi,

I’m sorry, I disagree – every single time you run AzCopy, you can incur a 1-second delay. Perhaps there was confusion about where the delay is. When I referred to a job, I didn't mean AzCopy generating thousands of plans in one pass and each job running a plan takes that time.

I was referring to the 4500 times I call AzCopy in my batch file job as per the MS recommendations https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-optimize

"One way to reduce the size of a job is to limit the number of files affected by a job. You can use command parameters to do that. For example, a job can copy only a subset of directories by using the include path parameter as part of the azcopy copy command."

e.g.
c:\az\azcopy.exe copy folder1 folder2
-> 1 second delay
c:\az\azcopy.exe copy folder3 folder4
-> 1 second delay
c:\az\azcopy.exe copy folder5 folder6
-> 1 second delay

Try repeating that 4500 times.

I’ve built my own copy already (to do the timing test, and fix that 2-second delay and yes, already removed the getversion) to overcome these arbitrarily enforced delays that ultimately cost me and my company time .. which is money.

Perhaps it doesn’t affect you sitting right near the version server, but the ability for a client to choose to stop this check via a flag is required for clients of azcopy to save a significant amount of time. It is their informed choice to turn off this check to save them time.

**Comment by mohsha-msft** [comment]
Hey @supermistag ,

I understand your problem. 
As you know AzCopy is an open source product, so we really encourage/appreciate users contributions. I see you've already removed these changes in your copy of build. That solves the problem I hope. I'm also curious to know about your findings in 2-sec delays and other improvements you may have thought in the process. We'll try to merge the fix of 2-sec delay in next release.
 
But, when you are trying to transfer as much as 18M files, the center of gravity lies in reducing the runtime of AzCopy sync command. Have you tried moving your machine (or VM) in same region as your storage account? Moving/creating VM in the same region as your storage account can significantly improve the performance by reducing scanning time and network transfer latencies. 

Also, you can run [AzCopy Benchmark](https://github.com/Azure/azure-storage-azcopy#supported-operations) command to find the correct tuning parameters specific to your environment. 

You can find all our documentation on [publicly available pages](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-v10) or in the [AzCopy Wiki Section](https://github.com/Azure/azure-storage-azcopy/wiki)

Please reach out if you have question about optimizations. I'll be more than happy to help 😃

**Comment by supermistag** [comment]
The data is stored on a SAN, so moving the source machine(s) is not an option. Believe me, I've gone over as much documentation on improving performance as I can find. Some of it is inconsistent to the source code.

I commented on the other 2-second issue saying the proposed fix works well.

**Comment by mohsha-msft** [comment]
I am thinking of optimizing the version check so it can cache the value and doesn't have to make REST Calls multiple times. I'll keep you posted :)

**Comment by supermistag** [comment]
[Optimize AzCopy](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-optimize) refers to check-length for both copy and sync. Check-length doesn't exist in sync (logged #1704, closed as ... i don't know why)

Tried using --include-after, but that was closed for azcopy in #1398 (didn't read the OP question, they asked for support in azcopy sync)

**Comment by Darryl-G** [comment]
> The data is stored on a SAN, so moving the source machine(s) is not an option. Believe me, I've gone over as much documentation on improving performance as I can find. Some of it is inconsistent to the source code.
> 
> I commented on the other 2-second issue saying the proposed fix works well.

You could block it on the source host(s) by adding a hosts file entry to 127.0.0.1 for host azcopyvnextrelease.blob.core.windows.net
This way the traffic would not even leave the vNIC.
Obviously it would prevent you easily updating AzCopy (could also be a positive thing to prevent the version being auto-updated).
Pros & cons I guess.

**Comment by vijaytdh** [comment]
In our case the we see the initial request to aka.ms which is dropped by Azure Firewall because we are not going to allow blanket access to this domain. It does mean that the job that periodically runs azcopy runs, it spams the log (in our case this runs every 1 min as its sync'ing  files from a storage account) files with errors around the check failing. It makes the logs noisy and it means any alerts we want to create off the back of this have to exclude this from the logs.

**Comment by malthe** [comment]
@Darryl-G I tried blocking it by adding a hosts file entry for azcopyvnextrelease.blob.core.windows.net to point to 127.0.0.1 – rather than make startup faster, it now takes a flat 8 seconds.

**Comment by Darryl-G** [comment]
> @Darryl-G I tried blocking it by adding a hosts file entry for azcopyvnextrelease.blob.core.windows.net to point to 127.0.0.1 – rather than make startup faster, it now takes a flat 8 seconds (consistently).
> 
> ```shell
> $ time azcopy -v
> real 0m8.014s
> ```
> 
> Previously, I get around 0.6s.

AzCopy must be trying and trying until it hits some sort of timeout.  Instead of 127.0.0.1  put some random IP that does not exist on your internal network.  See if that changes the behavior.

**Comment by malthe** [comment]
But presumably, what the program would do is the equivalent of `telnet 127.0.0.1 443`.

When I do that on the machine, I get:
```bash
$ telnet azcopyvnextrelease.blob.core.windows.net 443
Trying 127.0.0.1...
telnet: Unable to connect to remote host: Connection refused
```
This happens immediately.

The point is, `azcopy` must be doing a round (or several) of retrying on its own account. I don't think that can be worked around by picking a different IP.

**Comment by Darryl-G** [comment]
A "connection refused" error is different to a "cannot resolve" error.
Underlying TCP/IP, the network stack still resolves 127.0.0.1 to a valid
MAC address and will allow the connection attempt on port 443.
If it is unable to get a MAC, then you might find a different outcome.
I can't be sure without testing.  I'm just saying that it might produce a
different outcome because it is a different scenario.
Note: it needs to be either a local class C up address or a private one
"192.x" that you choose. To prevent it trying to resolve something on the
internet.
